syntax = "proto3";

package flightrecorder.v1;

option go_package = "github.com/lokryn-suite/flightrecorder/proto;pb";

// ---------- Severity ----------
enum Severity {
  SEVERITY_UNSPECIFIED = 0;
  SEVERITY_DEBUG       = 1;
  SEVERITY_INFO        = 2;
  SEVERITY_NOTICE      = 3;
  SEVERITY_WARNING     = 4;
  SEVERITY_ERROR       = 5;
  SEVERITY_CRITICAL    = 6;
  SEVERITY_ALERT       = 7;
  SEVERITY_EMERGENCY   = 8;
}

// ---------- Outcome ----------
enum Outcome {
  OUTCOME_UNSPECIFIED        = 0;
  OUTCOME_SUCCESS            = 1;
  OUTCOME_FAILURE_UNAUTHORIZED = 2;
  OUTCOME_FAILURE_DENIED     = 3;
  OUTCOME_FAILURE_ERROR      = 4;
  OUTCOME_PARTIAL            = 5;
}

// ---------- EventType ----------
enum EventType {
  EVENT_TYPE_UNSPECIFIED     = 0;
  EVENT_LOGIN                = 1;
  EVENT_LOGOUT               = 2;
  EVENT_FILE_ACCESS          = 3;
  EVENT_POLICY_CHANGE        = 4;
  EVENT_PRIVILEGE_USE        = 5;
  EVENT_CONFIG_CHANGE        = 6;
  EVENT_DATA_EXPORT          = 7;
  EVENT_NETWORK_CONNECTION   = 8;
  EVENT_PROCESS_START        = 9;
  EVENT_PROCESS_STOP         = 10;
  EVENT_USER_MANAGEMENT      = 11;
  EVENT_RESOURCE_ACCESS      = 12;
  EVENT_TOOL_INVOCATION      = 13;  // MCP tool call, function call, API invocation by agent
  EVENT_MODEL_INFERENCE      = 14;  // LLM API call (completion, embedding, etc.)
  EVENT_AGENT_DECISION       = 15;  // Autonomous decision point (branching, action selection)
  EVENT_DELEGATION           = 16;  // Agent delegating to sub-agent or another system
  EVENT_CONTEXT_ACCESS       = 17;  // Accessing context window, memory, RAG retrieval
  EVENT_PROMPT_EXECUTION     = 18;  // Prompt template execution
  EVENT_GUARDRAIL_CHECK      = 19;  // Safety/guardrail evaluation (pass/fail)
}


// ---------- Sensitivity ----------
enum Sensitivity {
  SENSITIVITY_UNSPECIFIED     = 0;
  SENSITIVITY_PUBLIC          = 1;
  SENSITIVITY_INTERNAL        = 2;
  SENSITIVITY_CONFIDENTIAL    = 3;
  SENSITIVITY_RESTRICTED      = 4;
  SENSITIVITY_HIGHLY_RESTRICTED = 5;
}

// ---------- LogRequest ----------
message LogRequest {
  EventType event_type   = 1;
  Outcome outcome        = 2;
  Severity severity      = 3;
  string actor_id        = 4;
  string component       = 5;
  string environment     = 6;
  string resource        = 7;
  string message         = 8; // free-form, human-readable
  bytes payload          = 9; // opaque JSON/CBOR blob
  repeated string policy_tags = 10;
  Sensitivity sensitivity = 11;
}

// ---------- Service API ----------
service FlightRecorder {
  rpc AppendLog(LogRequest) returns (AppendResponse);
  rpc StreamLogs(stream LogRequest) returns (StreamAck);
}

message AppendResponse {
  bool accepted = 1;
  string error  = 2;
}

message StreamAck {
  uint64 received_count = 1;
}